### PR TITLE
Retry expired watches

### LINKED ISSUE #57: Watch stream should handle HTTP error before unmarshaling event
I could be mistaken but looking at the infinite loop for the watch stream doesn't handle the case when you receive an event that is expired i.e. a HTTP status code of 410. 

``` python
        while True:
            resp = func(*args, **kwargs)
            try:
                for line in iter_resp_lines(resp):
                    yield self.unmarshal_event(line, return_type)
                    if self._stop:
                        break
            finally:
                kwargs['resource_version'] = self.resource_version
                resp.close()
                resp.release_conn()
```
Looking at the code it seems that if the event is expired then resp should return something along the lines of 
```
{'raw_object': {u'status': u'Failure', u'kind': u'Status', u'code': 410, u'apiVersion': u'v1', u'reason': u'Gone', u'message': u'too old resource version: 2428 (88826)', u'metadata': {}}, u'object': {'api_version': 'v1',
 'kind': 'Status',
```
And `unmarshall_event` should fail to deserialize the object and break. And `self.resource_version` should just be stuck on the resource_version of the event that was expired. 

Am I missing something here?

### PR DESCRIPTION
**Pull request by mitar**
This is a followup to #102 and fixes #57.

### CODE REVIEW

**Comment by ccamacho** [comment]
It would be nice to have those 3 commits squashed into 1

**Comment by mitar** [comment]
Doesn't GitHub support that when merging the PR?

**Comment by micw523** [comment]
Manually yes, but we’re using bots to merge the commits so usually the commits need to be squashed.

**Comment by mitar** [comment]
Done.

File: watch/watch.py [hunk 1]
```diff
@@ -138,11 +146,26 @@ def stream(self, func, *args, **kwargs):
...
         timeouts = ('timeout_seconds' in kwargs)
+        retry_after_410 = False
         while True:
             resp = func(*args, **kwargs)
             try:
                 for line in iter_resp_lines(resp):
-                    yield self.unmarshal_event(line, return_type)
+                    event = self.unmarshal_event(line, return_type)
+                    if isinstance(event, dict) and event['type'] == 'ERROR':
+                        obj = event['raw_object']
```
Comment (micw523) [inline watch/watch.py]: Is it safe to assume the APIs won't change for accessing ```event['raw_object']['code']```, @roycaihw ?

File: watch/watch.py [hunk 2]
```diff
@@ -138,11 +146,26 @@ def stream(self, func, *args, **kwargs):
...
+                        # Current request expired, let's retry,
+                        # but only if we have not already retried.
+                        if not retry_after_410 and obj['code'] == 410:
+                            retry_after_410 = True
+                            break
```
Comment (roycaihw) [inline watch/watch.py]: we should not retry the watch request with the same RV when it's already gone in apiserver

we can retry on 500 and 504 according to the [client-go implementation](https://github.com/kubernetes/kubernetes/blob/270d9529e004f04829a7cc9e5c83e43160a44f3a/staging/src/k8s.io/client-go/tools/watch/retrywatcher.go#L203-L211)

File: watch/watch.py [hunk 3]
```diff
@@ -91,7 +91,7 @@ def unmarshal_event(self, data, return_type):
         except ValueError:
             return data
         js['raw_object'] = js['object']
-        if return_type:
+        if return_type and js['type'] != 'ERROR':
```
Comment (roycaihw) [inline watch/watch.py]: this is where we fail to deserialize the response. Good catch

File: watch/watch.py (hunk 2 above)
Comment (roycaihw) [reply watch/watch.py]: @yliaog pointed out that we need to be careful about the code structure here. We want to be consistent to kubernetes client-go in terms of what a watch is capable of doing

- In client-go, a watch function call [doesn't retry on err](https://github.com/kubernetes/kubernetes/blob/5219ad7be209b31cdfdae2eda1cd435aff4a3cd2/staging/src/k8s.io/apimachinery/pkg/watch/streamwatcher.go#L99). A client can implement their own handling for watch errors
- client-go provides tooling like informer and retrywatcher for clients to leverage, which have retry logics for watch errors

We should be consistent and provide the same choices to our user

**Comment by dmateusp** [comment]
is this good to go ?

**Comment by jinchihe** [comment]
Can someone review and approve the PR? we meet the problem now... Thanks.

File: watch/watch.py (hunk 2 above)
Comment (mitar) [reply watch/watch.py]: But that would be backwards incompatible with current offering by Python package? It already supports retrying. So I think for Python it seems we already made decision that watch is retrying.

**Comment by mel1nn** [comment]
Could you fix the merge conflict please ?

**Comment by mel1nn** [comment]
@mitar ?

**Comment by mitar** [comment]
Sorry. I will look into this now.

**Comment by mitar** [comment]
Done.

**Comment by mitar** [comment]
I have updated this. Please review it before it gets rotten again.

**Comment by mel1nn** [comment]
@mbohlool ?

**Comment by mel1nn** [comment]
@roycaihw ?

File: watch/watch.py [hunk 4]
```diff
@@ -134,14 +142,30 @@ def stream(self, func, *args, **kwargs):
...
                     # return raw string when we are streaming log
                     if watch_arg == "watch":
-                        yield self.unmarshal_event(line, return_type)
+                        event = self.unmarshal_event(line, return_type)
+                        if isinstance(event, dict) \
+                                and event['type'] == 'ERROR':
+                            obj = event['raw_object']
+                            # Current request expired, let's retry,
+                            # but only if we have not already retried.
+                            if not retry_after_410 and obj['code'] == 410:
```
Comment (yliaog) [inline watch/watch.py:159]: better to use the code in http.HTTPStatus to avoid using magic const 410

Comment (mitar) [reply watch/watch.py]: Updated.

File: watch/watch.py [hunk 5]
```diff
@@ -102,6 +103,14 @@ def unmarshal_event(self, data, return_type):
     def stream(self, func, *args, **kwargs):
         """Watch an API resource and stream the result back via a generator.
 
+        Note that watching an API resource can expire. The method tries to
+        resume automatically from the last result, but if that last result
```
Comment (yliaog) [inline watch/watch.py:107]: i think this comment has already implied it, but still better to make it clear that only *retry once*, so user has the correct expectation on how many times retry is attempted.

**Comment by yliaog** [comment]
thanks for the PR.

/lgtm
/approve

**Pull request merged**